modules/rpc/yarpc.go

```diff
@@ -105,78 +196,85 @@ func newYarpcModule(
-	module.config = *cfg
+	err = module.Host().Config().Scope("modules").Get(module.Name()).PopulateStruct(&module.config)
+	if err != nil {
+		return module, err
+	}
+
+	module.config.inboundMiddleware = inboundMiddlewareFromCreateInfo(mi)
+	module.config.onewayInboundMiddleware = onewayInboundMiddlewareFromCreateInfo(mi)
 
-	module.inboundMiddlewares = inboundMiddlewaresFromCreateInfo(mi)
-	module.onewayInboundMiddlewares = onewayInboundMiddlewaresFromCreateInfo(mi)
+	tchTransport, err := tch.NewChannelTransport(
+		tch.ServiceName(module.config.AdvertiseName),
+		tch.ListenAddr(module.config.Bind),
+	)
+
+	module.config.inbounds = []transport.Inbound{tchTransport.NewInbound()}
 
 	return module, err
 }
 
-// Start begins serving requests over YARPC
-func (m *YarpcModule) Start(readyCh chan<- struct{}) <-chan error {
+// Start begins serving requests over RPC. It first stops the current dispatcher, merges configs from the global
+// collection and then starts the new dispatcher.
+func (m *YARPCModule) Start(readyCh chan<- struct{}) <-chan error {
+	if m.IsRunning() {
+		panic("module is already running")
+	}
+
 	m.stateMu.Lock()
 	defer m.stateMu.Unlock()
 
-	interceptor := yarpc.UnaryInboundMiddleware(m.inboundMiddlewares...)
-	onewayInterceptor := yarpc.OnewayInboundMiddleware(m.onewayInboundMiddlewares...)
-
 	ret := make(chan error, 1)
-	tchTransport, err := tch.NewChannelTransport(
-		tch.ServiceName(m.config.AdvertiseName),
-		tch.ListenAddr(m.config.Bind),
-	)
-	if err != nil {
-		ret <- errors.New("Unable to create TChannel transport " + err.Error())
+	if err := _configs.addConfig(&m.config); err != nil {
+		ret <- errs.Wrap(err, "can't add module config")
 		return ret
 	}
 
-	m.rpc, err = _dispatcherFn(m.Host(), yarpc.Config{
-		Name: m.config.AdvertiseName,
-		Inbounds: []transport.Inbound{
-			tchTransport.NewInbound(),
-		},
-		InboundMiddleware: yarpc.InboundMiddleware{
-			Unary:  interceptor,
-			Oneway: onewayInterceptor,
-		},
-		Tracer: m.Tracer(),
-	})
-
-	if err != nil {
+	if config, err := _configs.mergeConfigs(); err != nil {
+		ret <- errs.Wrap(err, "unable to merge configs from multiple modules")
+		return ret
+	} else if err := resetDispatcher(m.Host(), config); err != nil {
 		ret <- err
 		return ret
 	}
 
 	m.register(m)
-	// TODO update log object to be accessed via context.Context #74
-	m.log.Info("Service started", "service", m.config.AdvertiseName, "port", m.config.Bind)
+	m.Host().Logger().Info("Module started",
+		"name", m.config.AdvertiseName,
+		"port", m.config.Bind)
 
-	ret <- m.rpc.Start()
+	m.isRunning = true
 	readyCh <- struct{}{}
 	return ret
 }
 
-// Stop shuts down a YARPC module
-func (m *YarpcModule) Stop() error {
+// Stop shuts down a YARPC module. It will stop the current dispatcher, remove the module config from the
+// config collection and start the new dispatcher with a new merged config.
+func (m *YARPCModule) Stop() error {
+	if !m.IsRunning() {
+		panic("module is not running")
```

Comment:
This will only happen once we're shutting down the whole application right?  If not it seems unsafe to have a panic which could damage the whole application.  We might want to consider making start/stop idempotent.